### PR TITLE
Collapse with flex elements

### PR DESCRIPTION
**Pull request by Johann-S**
Detect if the element use flexbox, if it's the case add `.flexshow`

- [x] Add Visual test
- [x] Detect all posible flex values (inline-flex, flex, etc...)

We detect flex by creating hidden div(s) and we add each class to this hidden div to detect if this div is displayed in flex mode or not. 

We do that only when we call the constructor of Collapse plugins that's why I added an `update` method if the collapsible element change, to detect once again the appropriate show class

Close : #22600

/CC : @mdo

### CODE REVIEW

File: js/src/collapse.js [hunk 1]
```diff
@@ -298,6 +309,31 @@ const Collapse = (($) => {
       }
     }
 
+    _getShowClass() {
+      const tabClass = this._element.classList
+
+      // Detect flex for inline style
+      let useFlex = $(this._element).css('display').indexOf('flex') !== -1
+
+      // Create a wrapper to hide our flex detection
```
Comment (glebm) [inline js/src/collapse.js]: Can you not use `getComputedStyle(this._element).display` instead?

Comment (Johann-S) [inline js/src/collapse.js]: I can use that for the line 316 yes, but not for the detection below because getComputedStyle return the current applied style and that's not what I want here. 
I want to know if they are a class which contain a flex display

**Comment by mdo** [comment]
Is anything here considered to be a breaking change @Johann-S? Might be beneficial to wait for Beta 2 to address this.

**Comment by Johann-S** [comment]
That's right this can wait for beta 2

File: js/src/collapse.js [hunk 2]
```diff
@@ -314,6 +325,31 @@ const Collapse = (($) => {
       }
     }
 
+    _getShowClass() {
+      const tabClass = this._element.classList
+
+      // Detect flex for inline style
+      let useFlex = $(this._element).css('display').indexOf('flex') !== -1
```
Comment (XhmikosR) [inline js/src/collapse.js]: How about just using https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_Objects/String/includes?

File: js/tests/visual/collapse.html [hunk 3]
```diff
@@ -55,9 +55,25 @@ <h5 class="mb-0">
...
+                Anim pariatur cliche reprehenderit, enim eiusmod high life accusamus terry richardson ad squid. Nihil anim keffiyeh helvetica, craft beer labore wes anderson cred nesciunt sapiente ea proident.
+              </div>
+            </div>
+          </div>
+        </div>
+      </div>
     </div>
 
-    <script src="../../../docs/assets/js/vendor/jquery-slim.min.js"></script>
+    <script src="../../../assets/js/vendor/jquery-slim.min.js"></script>
```
Comment (XhmikosR) [inline js/tests/visual/collapse.html]: Hmm, how did it work without this?

File: docs/4.0/components/collapse.md [hunk 4]
```diff
@@ -226,6 +226,10 @@ Shows a collapsible element. **Returns to the caller before the collapsible elem
 
 Hides a collapsible element. **Returns to the caller before the collapsible element has actually been hidden** (i.e. before the `hidden.bs.collapse` event occurs).
 
+#### `.collapse('update')`
+
+Update a collapsible element if this element changed during his lifetime
```
Comment (XhmikosR) [inline docs/4.0/components/collapse.md]: `its lifetime`

Comment (XhmikosR) [inline docs/4.0/components/collapse.md]: And maybe add a `.` at the end.

File: js/tests/visual/collapse.html (hunk 3 above)
Comment (Johann-S) [inline js/tests/visual/collapse.html]: It didn't work before that change and yes it's an unrelated change

File: js/src/collapse.js (hunk 2 above)
Comment (Johann-S) [inline js/src/collapse.js]: I switched to `includes` but they are no support on IE 10 so back to `indexOf`

Comment (XhmikosR) [inline js/src/collapse.js]: But it's transpiled isn't it?

File: docs/4.0/components/collapse.md [hunk 5]
```diff
@@ -226,6 +226,10 @@ Shows a collapsible element. **Returns to the caller before the collapsible elem
 
 Hides a collapsible element. **Returns to the caller before the collapsible element has actually been hidden** (i.e. before the `hidden.bs.collapse` event occurs).
 
+#### `.collapse('update')`
+
+Update a collapsible element if this element changed during its lifetime.
```
Comment (XhmikosR) [inline docs/4.0/components/collapse.md]: How about: "Update a collapsible element if it changed during its lifetime."

File: js/src/collapse.js (hunk 2 above)
Comment (Johann-S) [inline js/src/collapse.js]: I didn't checked but unit tests failed

File: docs/4.0/components/collapse.md (hunk 5 above)
Comment (Johann-S) [inline docs/4.0/components/collapse.md]: Yep that's better thanks 👍 I'll update that soon

File: js/src/collapse.js (hunk 2 above)
Comment (XhmikosR) [inline js/src/collapse.js]: But since we use babel it should be taken care of. Please try one more time.

Comment (XhmikosR) [inline js/src/collapse.js]: I looked into it a bit more and it seems Babel doesn't do this. Just verify one more time and if it doesn't just use `indexOf`. It's not worth using a polyfill for this.

Comment (XhmikosR) [inline js/src/collapse.js]: Just keep in mind we can do this if we just load the babel polyfill and use ES6 methods. Not sure what the exact size diff is, just thinking out loud for another PR.

Comment (Johann-S) [inline js/src/collapse.js]: Do you know which polyfill to use ? I can make a test 👍

Comment (XhmikosR) [inline js/src/collapse.js]: https://babeljs.io/docs/usage/polyfill/

**Comment by Johann-S** [comment]
For me everything is fine @mdo so this one can be merged 👍

File: js/src/collapse.js (hunk 2 above)
Comment (Johann-S) [reply js/src/collapse.js]: I checked `babel-polyfill` and currently when we import babel-polyfill we will import all the available polyfill not the `includes` one so for now we should merge this PR and open an issue on how to integrate babel-polyfill with only specific es6 polyfill

Comment (XhmikosR) [reply js/src/collapse.js]: No worries, it's not so important currently as long as we don't use features that aren't available for the browsers we support.

**Comment by vanillajonathan** [comment]
Keep in mind that the other `.flex-*` classes are separated with a hyphen.
Example `.flow-row`, `.flex-column`, `.flex-wrap`, etc.'.

Perhaps `.flex-show` should be considered.

File: scss/_transitions.scss [hunk 6]
```diff
@@ -14,16 +14,21 @@
...
   }
+  &.flexshow {
+    display: flex;
+  }
 }
 
 tr {
-  &.collapse.show {
+  &.collapse.show,
+  &.collapse.flexshow {
```
Comment (vanillajonathan) [inline scss/_transitions.scss]: Should `.flexshow` be named `.flex-show` to be consistent with the hyphen separation used by the other `.flex-*` classes?

**Comment by Johann-S** [comment]
Not sure to keep that PR, because I have to create an hidden div to detect if folks use `flex` attributes or not, a better solution would be CSS one
